docs/tutorials/cooperative_groups_tutorial.rst

```diff
@@ -0,0 +1,312 @@
+length sequences and sum of ``result_size``/ ``BlockSize`` length sequences. The
+host side reference implementation is the following:
+
+.. code-block:: cpp
+
+  // Host side function to perform the same reductions as executed on the GPU
+  std::vector<unsigned int> ref_reduced(const unsigned int        partition_size,
+                                        std::vector<unsigned int> input)
+  {
+      const unsigned int        input_size  = input.size();
+      const unsigned int        result_size = input_size / partition_size;
+      std::vector<unsigned int> result(result_size);
+
+      for(unsigned int i = 0; i < result_size; i++)
+      {
+          unsigned int partition_result = 0;
+          for(unsigned int j = 0; j < partition_size; j++)
+          {
+              partition_result += input[partition_size * i + j];
+          }
+          result[i] = partition_result;
+      }
+
+      return result;
+  }
+
+Device side code
+----------------
+
+.. TODO: Add link here to reduction tutorial and subsections
+
+To be able to calculate the sum of the sets of numbers, the tutorial using the
+shared memory based reduction at device side. In this tutorial the warp level
+intrinsics usage not covered like in the reduction tutorial. The ``x`` input
+variable is a shared pointer, which needs to be synchronized after every value
+changes. The ``thread_group`` input parameter can be ``thread_block_tile`` or
+``thread_block`` also, because the ``thread_group`` is the parent class of these
+types. The ``val`` is the numbers, which we wants to calculate the sum of. The
+returned results of this function return the final results of the reduction on
+thread ID 0 of the ``thread_group`` and at every other threads the function
+results are 0.
+
+.. code-block:: cpp
+
+  /// \brief Summation of `unsigned int val`'s in `thread_group g` using shared memory `x`
+  __device__ unsigned int reduce_sum(thread_group g, unsigned int* x, unsigned int val)
+  {
+      // Rank of this thread in the group
+      const unsigned int group_thread_id = g.thread_rank();
+
+      // We start with half the group size as active threads
+      // Every iteration the number of active threads halves, until we processed all values
+      for(unsigned int i = g.size() / 2; i > 0; i /= 2)
+      {
+          // Store value for this thread in a shared, temporary array
+          x[group_thread_id] = val;
+
+          // Synchronize all threads in the group
+          g.sync();
+
+          // If our thread is still active, sum with its counterpart in the other half
+          if(group_thread_id < i)
+          {
+              val += x[group_thread_id + i];
+          }
+
+          // Synchronize all threads in the group
+          g.sync();
+      }
+
+      // Only the first thread returns a valid value
+      if(g.thread_rank() == 0)
+          return val;
+      else
+          return 0;
+  }
+
+The ``reduce_sum`` device function reused to calculate the block and custom
```

Comment:
reused?

Reply:
What is the problem with the reused, word?